src/Stratis.Bitcoin/Consensus/PeerGetHeaderRequestBehavior.cs

```diff
@@ -0,0 +1,132 @@
+        /// <summary>How long the offending peer will be banned for.</summary>
+        private const int BanDurationSeconds = 3600;
+
+        /// <summary>Provider of time functions.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>The last header that was requested.</summary>
+        private uint256 getHeaderLastRequestHash;
+
+        /// <summary>The last time the header was requested.</summary>
+        private DateTimeOffset? getHeaderLastRequestedTimestamp;
+
+        /// <summary>The amount of times the same request has been made.</summary>
+        private int getHeaderRequestCount;
+
+        /// <summary>The threshold after which the node will be banned and disconnected.</summary>
+        private const int GetHeaderRequestCountThreshold = 10;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Factory for creating loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Contract for peer banning behavior.</summary>
+        private readonly IPeerBanning peerBanning;
+
+        public RateLimitingBehavior(
+            IDateTimeProvider dateTimeProvider,
+            ILoggerFactory loggerFactory,
+            IPeerBanning peerBanning)
+        {
+            this.dateTimeProvider = dateTimeProvider;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.peerBanning = peerBanning;
+        }
+
+        public override object Clone()
+        {
+            return new RateLimitingBehavior(this.dateTimeProvider, this.loggerFactory, this.peerBanning);
+        }
+
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceived, true);
+        }
+
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceived);
+        }
+
+        /// <summary>
+        /// Processes an incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private Task OnMessageReceived(INetworkPeer peer, IncomingMessage message)
+        {
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    this.AssessPeerGetHeaderFrequency(getHeaders);
+                    break;
+
+                default:
+                    break;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AssessPeerGetHeaderFrequency(GetHeadersPayload getHeaders)
+        {
+            // Is the last requested hash the same as this request.
+            if (this.getHeaderLastRequestHash == getHeaders.BlockLocator.Blocks.First())
```

Comment:
This will throw if block locator is empty

Reply:
Will Fix.

Reply:
Can it be empty?

Reply:
yes, no prior validation was done at the point the payload is processed here.